File: C++_Version/String/345Reverse_Vowels_of_a_String.cpp
//Write a function that takes a string as input and reverse only the vowels of a string.
//
//Example 1:
//Given s = "hello", return "holle".
//
//Example 2 :
//	Given s = "leetcode", return "leotcede".

#include <vector>
#include <iostream>
#include <algorithm>
#include <cstdlib>
#include <functional>
#include <string>
#include <unordered_map>
#include <map>
#include <set>
#include <unordered_set>
#include <cctype>
using namespace std;

class Solution {
public:
	string reverseVowels(string s) {
		size_t s_length = s.length();

		if (s_length <= 1)
			return s;

		unordered_set<char> vowels;
		vowels.insert('a');
		vowels.insert('e');
		vowels.insert('i');
		vowels.insert('o');
		vowels.insert('u');
		vowels.insert('A');
		vowels.insert('E');
		vowels.insert('I');
		vowels.insert('O');
		vowels.insert('U');

		
		int l_p = 0, h_p = s_length - 1;

		while (l_p < h_p)
		{
			while (l_p < h_p && vowels.find(s[l_p]) == vowels.end())
				l_p++;
 			while (l_p < h_p && vowels.find(s[h_p]) == vowels.end())
				h_p--;

			char temp = s[l_p];
			s[l_p] = s[h_p];
			s[h_p] = temp;

			l_p++;
			h_p--;
		}

		return s;
	}
};

int main()
{
	Solution sol1;

	string A("a."), B("baaabbbababaaabbbababaaabbbababaaabbbababaaabbbababaaabbbababaaabbbababaaabbbababaaabbbababaaabbbaba");

	string receive = sol1.reverseVowels(A);

	system("pause");

	return 0;
}

File: Java_Version/345_Reverse_Vowels_of_a_String.java
/*
  Write a function that takes a string as input and reverse only the vowels of
a string.

Example 1:

        Given s = "hello", return "holle".

Example 2:

        Given s = "leetcode", return "leotcede".

Note:
  The vowels does not include the letter "y".
*/

class Solution {

    public static void main(String[] args) {

        System.out.println(new Solution().reverseVowels("lEetcode"));
    }

    public String reverseVowels(String s) {

        if (s == null || s.length() == 0) {
            return s;
        }

        char[] chars = s.toCharArray();
        int startPointer = 0, endPointer = chars.length - 1;

        while (startPointer < endPointer) {

            while (startPointer < endPointer && !isVowelChar(chars[startPointer])) {
                startPointer++;
            }
            while (startPointer < endPointer && !isVowelChar(chars[endPointer])) {
                endPointer--;
            }

            if (startPointer < endPointer) {
                swap(chars, startPointer, endPointer);
                startPointer++;
                endPointer--;
            }
        }

        return new String(chars);
    }

    private boolean isVowelChar(char c) {

        return c == 'a' || c == 'e' || c == 'i' || c == 'o' || c == 'u' ||
                c == 'A' || c == 'E' || c == 'I' || c == 'O' || c == 'U';
    }

    private void swap(char[] chars, int index1, int index2) {

        char temp = chars[index1];

        chars[index1] = chars[index2];
        chars[index2] = temp;
    }
}
